rdkPlugins/DeviceMapper/source/DeviceMapper.cpp

```diff
@@ -0,0 +1,233 @@
+/*
+* If not stated otherwise in this file or this component's LICENSE file the
+* following copyright and licenses apply:
+*
+* Copyright 2021 Sky UK
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+#include "DeviceMapper.h"
+
+#include <Logging.h>
+#include <algorithm>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/sysmacros.h>
+#include <unistd.h>
+
+
+REGISTER_RDK_PLUGIN(DeviceMapperPlugin);
+
+DeviceMapperPlugin::DeviceMapperPlugin(std::shared_ptr<rt_dobby_schema> &containerConfig,
+                                       const std::shared_ptr<DobbyRdkPluginUtils> &utils,
+                                       const std::string &rootfsPath)
+    : mName("DeviceMapper"),
+    mContainerConfig(containerConfig),
+    mRootfsPath(rootfsPath),
+    mUtils(utils)
+{
+    AI_LOG_FN_ENTRY();
+
+    if (!mContainerConfig || !mContainerConfig->rdk_plugins->devicemapper || !mContainerConfig->rdk_plugins->devicemapper->data)
+    {
+        mValid = false;
+    }
+    else
+    {
+        mValid = true;
+    }
+
+    AI_LOG_FN_EXIT();
+}
+
+/**
+ * @brief Set the bit flags for which hooks we're going to use
+ */
+unsigned DeviceMapperPlugin::hookHints() const
+{
+    return (IDobbyRdkPlugin::HintFlags::PreCreationFlag);
+}
+
+/**
+ * @brief Dobby Hook - run in host namespace before container creation process
+ */
+bool DeviceMapperPlugin::preCreation()
+{
+    AI_LOG_FN_ENTRY();
+
+    if (!mValid)
+    {
+        AI_LOG_WARN("Invalid config file");
+        return false;
+    }
+
+    if (mContainerConfig->rdk_plugins->devicemapper->data->devices_len == 0 || mContainerConfig->linux->devices_len == 0)
+    {
+        // No devices in plugin config or in container config, nothing to do
+        return true;
+    }
+
+    // Get the major/minor IDs of the devices we're interested in
+    std::vector<DevNode> deviceNodes;
+    DevNode node;
+    for (size_t i = 0; i < mContainerConfig->rdk_plugins->devicemapper->data->devices_len; i++)
+    {
+        if (getDevNodeFromPath(mContainerConfig->rdk_plugins->devicemapper->data->devices[i], node))
+        {
+            deviceNodes.emplace_back(node);
+        }
+    }
+
+    // We know we have some devices we're interested in, but we didn't find any
+    // of them, so something's gone wrong
+    if (deviceNodes.size() == 0)
```

Comment:
So why aren't we also comparing 
`deviceNodes.size() != mContainerConfig->rdk_plugins->devicemapper->data->devices_len`?
Shouldn't them being not equal means that something went wrong as well? I know it gets logged as warning when some device was inaccessible which is good, but if it is fine when at least one device was mapped we should probably write some comment in README that this is the case. Or maybe we could add additional flag like "mandatory" which if true would mean that all devices must be mapped, and set it default false.